File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

#include <map>

namespace protein_translation
{
    const std::map<std::string, std::string> codon_to_protein = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},
        {"UAG", "STOP"},
        {"UGA", "STOP"},
    };

    std::vector<std::string> proteins(const std::string &rna)
    {
        std::vector<std::string> proteins{};
        std::string codon_buffer = "";

        for (char c : rna)
        {
            codon_buffer.push_back(c);
            if (codon_buffer.length() < 3)
                continue;

            std::string protein = codon_to_protein.at(codon_buffer);
            codon_buffer.clear();

            if (protein == "STOP")
            {
                break;
            }
            else
            {
                proteins.push_back(protein);
            }
        }

        return proteins;
    }

} // namespace protein_translation

File: solutions/haskell/protein-translation/1/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

fromJust :: Maybe a -> a
fromJust Nothing = error "Maybe.fromJust: Nothing"
fromJust (Just x) = x

proteins :: String -> Maybe [String]
proteins [] = Just []
-- Stop Codons
proteins ('U':'A':'A':x) = Just []
proteins ('U':'A':'G':x) = Just []
proteins ('U':'G':'A':x) = Just []
-- Other Codons
proteins ('A':'U':'G':x) = Just ("Methionine" : fromJust(proteins x))
proteins ('U':'U':'U':x) = Just ("Phenylalanine" : fromJust(proteins x))
proteins ('U':'U':'C':x) = Just ("Phenylalanine" : fromJust(proteins x))
proteins ('U':'U':'A':x) = Just ("Leucine" : fromJust(proteins x))
proteins ('U':'U':'G':x) = Just ("Leucine" : fromJust(proteins x))
proteins ('U':'C':'U':x) = Just ("Serine" : fromJust(proteins x))
proteins ('U':'C':'C':x) = Just ("Serine" : fromJust(proteins x))
proteins ('U':'C':'A':x) = Just ("Serine" : fromJust(proteins x))
proteins ('U':'C':'G':x) = Just ("Serine" : fromJust(proteins x))
proteins ('U':'A':'U':x) = Just ("Tyrosine" : fromJust(proteins x))
proteins ('U':'A':'C':x) = Just ("Tyrosine" : fromJust(proteins x))
proteins ('U':'G':'U':x) = Just ("Cysteine" : fromJust(proteins x))
proteins ('U':'G':'C':x) = Just ("Cysteine" : fromJust(proteins x))
proteins ('U':'G':'G':x) = Just ("Tryptophan" : fromJust(proteins x))
proteins _ = Nothing